apps/api/src/service/application-service.ts

```diff
@@ -0,0 +1,167 @@
+ * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+ * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import { and, asc, desc, eq, sql } from 'drizzle-orm';
+import { ApplicationStates } from 'pcgl-daco/packages/data-model/src/types.ts';
+import { type PostgresDb } from '../db/index.ts';
+import { applicationContents } from '../db/schemas/applicationContents.ts';
+import { applications } from '../db/schemas/applications.ts';
+
+const applicationService = (db: PostgresDb) => ({
+	createApplication: async ({ user_id }: { user_id: string }) => {
+		const newApplication: typeof applications.$inferInsert = {
+			user_id,
+			state: ApplicationStates.DRAFT,
+		};
+
+		try {
+			const newApplicationRecord = await db.insert(applications).values(newApplication).returning();
+			const { id } = newApplicationRecord[0];
+
+			const newAppContents: typeof applicationContents.$inferInsert = {
+				application_id: id,
+				created_at: new Date(),
+				updated_at: new Date(),
+			};
+			const newAppContentsRecord = await db.insert(applicationContents).values(newAppContents).returning();
+			const { id: contentsId } = newAppContentsRecord[0];
+
+			const application = await db
+				.update(applications)
+				.set({ contents: contentsId })
+				.where(eq(applications.id, id))
+				.returning();
+
+			return application[0];
+		} catch (err) {
+			console.error(`Error at createApplication with user_id: ${user_id}`);
+			console.error(err);
+			return null;
+		}
+	},
+	findOneAndUpdate: async ({ id, update }: { id: number; update: any }) => {
+		try {
+			const application = await db
+				.update(applications)
+				.set({ ...update, updated_at: sql`NOW()` })
+				.where(eq(applications.id, id))
+				.returning();
+
+			return application;
+		} catch (err) {
+			console.error(`Error at findOneAndUpdate with id: ${id}`);
+			console.error(err);
+			return null;
+		}
+	},
+	getApplicationById: async ({ id }: { id: number }) => {
+		try {
+			const applicationRecord = await db
+				.select()
+				.from(applications)
+				.where(eq(applications.id, id))
+				.leftJoin(applicationContents, eq(applications.contents, applicationContents.id));
+
+			const application = {
+				...applicationRecord[0].applications,
+				contents: applicationRecord[0].application_contents,
+			};
+
+			return application;
+		} catch (err) {
+			console.error(`Error at getApplicationById with id: ${id}`);
+			console.error(err);
+			return null;
+		}
+	},
+	listApplications: async ({
+		user_id,
+		state,
+		sort = '',
+		page = 0,
+		pageSize = 20,
+	}: {
+		user_id?: string;
+		state?: ApplicationStates;
+		sort?: string;
+		page?: number;
+		pageSize?: number;
+	}) => {
+		const isDescending = sort?.charAt(0) === '-';
+		const sortValue = isDescending ? sort.substring(1) : sort;
```

Comment:
This encodes a lot of rules for the content of the "sort" string value, and our type system won't validate if they are formatted correctly or contain valid fields for our data model. If we do use a string like this, we should add TSDoc comments on this function to ensure the content of this parameter are well documented. I have an alternate suggestion though:

I see the `-columnName` pattern for sort parameters used when encoded in URL query params, where you want to put condensed information and only string values. We don't have these restriction on our internal functions parameters. We can explicitly require our sort paramater to have two properties, one for direction and the other for the column name, and the values for these properties can be strictly set as `asc | desc` and as the list of legal column names for the schema you are sorting.

It would look something like:

```ts
type OrderBy<T extends string> = {
	direction: 'asc' | 'desc';
	column: T;
};
type ApplicationsColumnName = keyof typeof applications.$inferSelect;

// the sort function argument can be typed as:
(sort?: Array<OrderBy<ApplicationsColumnName>>) => { 

// Making this an array allows the function to orderBy multiple properties, sorting by user ascending and then created_at descending, for instance.

// The orderBy clause can be built like:
const orderByArguments = sort
	? sort.map((sortBy) =>
			sortBy.direction === 'asc' ? asc(applications[sortBy.column]) : desc(applications[sortBy.column]),
	  )
	: [asc(applications.created_at)];

	await db.select({}).orderBy(...orderByArguments);
}
```

Lastly, if we expect to do db sorting in multiple places, then we could build some of this code as common utilities, for sort types and to built the orderby arguments.

Reply:
Also added types/utils files